task/control-plane-agent/src/mgs_gimlet/host_phase2.rs

```diff
@@ -0,0 +1,285 @@
+                current.state = State::WaitingForFirstMgs {
+                    port,
+                    deadline: now + DELAY_TRY_OTHER_MGS,
+                };
+                port
+            }
+            State::WaitingForFirstMgs { port, deadline } => {
+                if now < deadline {
+                    return None;
+                }
+                // Timed out waiting for a response from the first MGS we tried;
+                // flip to the other one.
+                let port = match port {
+                    SpPort::One => SpPort::Two,
+                    SpPort::Two => SpPort::One,
+                };
+                current.state = State::WaitingForSecondMgs {
+                    port,
+                    deadline: now + DELAY_RETRY,
+                };
+                port
+            }
+            State::WaitingForSecondMgs { port, deadline } => {
+                if now < deadline {
+                    return None;
+                }
+                // Timed out waiting for a response from the second MGS we
+                // tried; flip back to the first and retry.
+                current.retry_count += 1;
+                if current.retry_count >= MAX_ATTEMPTS {
+                    current.notify_calling_task();
+                    self.current = None;
+                    return None;
+                }
+                let port = match port {
+                    SpPort::One => SpPort::Two,
+                    SpPort::Two => SpPort::One,
+                };
+                current.state = State::WaitingForFirstMgs {
+                    port,
+                    deadline: now + DELAY_TRY_OTHER_MGS,
+                };
+                port
+            }
+            State::Fetched => return None,
+        };
+
+        let message = Message {
+            header: Header {
+                version: gateway_messages::version::V2,
+                message_id,
+            },
+            kind: MessageKind::SpRequest(SpRequest::HostPhase2Data {
+                hash: current.hash,
+                offset: current.offset,
+            }),
+        };
+
+        let n = gateway_messages::serialize(tx_buf, &message).unwrap_lite();
+
+        Some(UdpMetadata {
+            addr: SP_TO_MGS_MULTICAST_ADDR,
+            port: MGS_UDP_PORT,
+            size: n as u32,
+            vid: vlan_id_from_sp_port(port),
+        })
+    }
+
+    pub(crate) fn ingest_incoming_data(
+        &mut self,
+        port: SpPort,
+        hash: [u8; 32],
+        offset: u64,
+        data: &[u8],
+    ) {
+        let current = match self.current.as_mut() {
+            Some(current)
+                if hash == current.hash && offset == current.offset =>
+            {
+                current
+            }
+            _ => return,
+        };
+
+        self.buffer.clear();
+        let n = usize::min(self.buffer.capacity(), data.len());
+        self.buffer.extend_from_slice(&data[..n]).unwrap_lite();
```

Comment:
Is it okay to discard the tail of `data` here?

Reply:
It shouldn't be possible to lose anything from `data` here, assuming it came from an MGS-sent packet (which technically this function doesn't know for sure, but in practice is true). `self.buffer` is sized large enough to hold a full UDP packet worth of data plus a bit, since packets all have some kind of metadata at the front but `buffer` is only given the payload. We could change this to assert that `data.len() <= self.buffer.capacity()`.

However, it's also _okay_ if we did end up losing data here somehow; the host requests for data only include the offset where the data should start, not a length. If we ended up dropping the tail of data here, we would have wasted the bandwidth spent to transfer that data, but nothing else would break. Something like:

1. Host requests data at offset 0
2. MGS sends us 1000 bytes starting at offset 0
3. Somehow `self.buffer` is undersized, so we only copy in 900 bytes
4. We give 900 bytes to the host
5. Host requests data at offset 900

would cause MGS to send the data we dropped (bytes 900-1000) twice but would otherwise function correctly.

Reply:
Added a comment explaining this